### PR TITLE
dockerfile for Alpine Linux support

### LINKED ISSUE #2403: Add dev_tools/docker/alpine/Dockerfile
See https://github.com/quantumlib/Cirq/issues/2289 for context.

### PR DESCRIPTION
**Pull request by jitendrs**
Fixes #2403

### CODE REVIEW

File: dev_tools/docker/apline/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    python3 -m pip install --upgrade pip && \
+    pip3 install --no-cache --upgrade pip setuptools wheel
+
+# Make python3 default
+RUN rm -f /usr/bin/python && ln -s /usr/bin/python3 /usr/bin/python
+
+# Install Cirq library from Github.
+RUN git clone https://github.com/quantumlib/Cirq
+# Install the needed Python libraries.
+RUN pip3 install -r Cirq/requirements.txt -r Cirq/cirq/contrib/contrib-requirements.txt -r Cirq/dev_tools/conf/pip-list-dev-tools.txt
```
Comment (Strilanc) [inline dev_tools/docker/apline/Dockerfile]: `python -m pip` is more reliable than `pip3` in my experience, though I guess in the context of a dockerfile you know it would keep working.

File: dev_tools/docker/apline/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+# Configure UTF-8 encoding.
+RUN sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && locale-gen
+ENV LANG en_US.UTF-8  
+ENV LANGUAGE en_US:en  
+ENV LC_ALL en_US.UTF-8 
+
+WORKDIR /Cirq
+
+EXPOSE 8888
```
Comment (vtomole) [inline dev_tools/docker/apline/Dockerfile]: Add newline

File: dev_tools/docker/apline/Dockerfile (hunk 1 above)
Comment (jitendrs) [reply dev_tools/docker/apline/Dockerfile]: I agree with you since other Docker file in our project was following pip3 command installation. I was trying to make a generic as it was maintained in other Docker file.
I would suggest implementing some changes to other docker files as well @Strilanc

File: dev_tools/docker/apline/Dockerfile (hunk 2 above)
Comment (jitendrs) [reply dev_tools/docker/apline/Dockerfile]: I am changing the installation command. This will fix this error.

**Comment by vtomole** [comment]
@jitendrs I see. Okay, try it out.

**Comment by jitendrs** [comment]
@vtomole In order to do so, I may need to remove tensorflow from runtime requirement. But that will lead to another problem for other Docker files. 
what do you suggest here? I am think to place require library at one place each of this requirement.txt will be categorized by type OS, for example, 
alpine_requirement.txt
ubantu_requirement.txt 

This can also resolved any installation issue coming in the near future. 
Let me know what do you think.

**Comment by vtomole** [comment]
@jitendrs Try this

https://github.com/better/alpine-tensorflow

**Comment by jitendrs** [comment]
@vtomole This example doesn't work with tensorflow 1.14. I will investigate and install it manually.

**Comment by Strilanc** [comment]
> I am think to place require library at one place each of this requirement.txt will be categorized by type OS, for example,
> alpine_requirement.txt
> ubantu_requirement.txt

I would **strongly** prefer to only have one set of python requirements.

Is there any version of tensorflow that installs on your system? We don't necessarily have to pin to the version we have.

**Comment by jitendrs** [comment]
@Strilanc I agree with you on python dependencies but alpine docker build failed to install TensorFlow with pip command. see below,

```
RUN python3 -m pip install tensorflow~=1.4
---> Running in 0bf137353ede
ERROR: Could not find a version that satisfies the requirement tensorflow~=1.14 (from versions: none)
ERROR: No matching distribution found for tensorflow~=1.14
```

Generic installation steps from the requirement files are failing. 
The only way I figure out is a manual compilation of TensorFlow with Bazel inside of docker file .

**Comment by vtomole** [comment]
@jitendrs We removed tensorflow as a dependency. Try it again.

**Comment by vtomole** [comment]
This is what i get

```
vtomole@vtomole:~/Cirq/dev_tools/docker/apline$ docker build .
Sending build context to Docker daemon   2.56kB
Step 1/11 : FROM alpine:latest
latest: Pulling from library/alpine
89d9c30c1d48: Pull complete 
Digest: sha256:c19173c5ada610a5989151111163d28a67368362762534d8a8121ce95cf2bd5a
Status: Downloaded newer image for alpine:latest
 ---> 965ea09ff2eb
Step 2/11 : RUN apk upgrade --update &&     apk add --no-cache bash git emacs vim &&     apk add libpng-dev freetype-dev openblas-dev libc-dev musl-dev &&     apk add python3-dev python3-tkinter &&     python3 -m pip install --upgrade pip &&     pip3 install --no-cache --upgrade pip setuptools wheel
 ---> Running in 6e55cac904e8
ERROR: Unable to read database state: No such file or directory
ERROR: Failed to open apk database: No such file or directory
The command '/bin/sh -c apk upgrade --update &&     apk add --no-cache bash git emacs vim &&     apk add libpng-dev freetype-dev openblas-dev libc-dev musl-dev &&     apk add python3-dev python3-tkinter &&     python3 -m pip install --upgrade pip &&     pip3 install --no-cache --upgrade pip setuptools wheel' returned a non-zero code: 99
```

File: dev_tools/docker/apline/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+RUN rm -f /usr/bin/python && ln -s /usr/bin/python3 /usr/bin/python
+
+# Install Cirq library from Github.
+RUN git clone https://github.com/quantumlib/Cirq
+# Install the needed Python libraries.
+RUN pip3 install -r Cirq/requirements.txt -r Cirq/cirq/contrib/contrib-requirements.txt -r Cirq/dev_tools/conf/pip-list-dev-tools.txt
+
+
+# Configure UTF-8 encoding.
+RUN sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && locale-gen
```
Comment (mpharrigan) [inline dev_tools/docker/apline/Dockerfile]: why is this necessary? put this at the top for more efficient cacheing of intermediate docker images

Comment (jitendrs) [reply dev_tools/docker/apline/Dockerfile]: @mpharrigan sure I will change it. 
 Right now there is another issue with qiskit-aer installation on alpine. 
https://github.com/Qiskit/qiskit-aer/issues/489

once I get answer I will update my PR.
Thanks
